apps/prairielearn/src/lib/question-render.ts

```diff
@@ -586,27 +640,37 @@ export async function renderPanelsForSubmission({
           instance_question_id: variant.instance_question_id,
         });
 
-  /** @type {SubmissionPanels} */
-  const panels = {
+  const group_role_permissions = await run(async () => {
+    if (!instance_question || !assessment_instance?.group_id || !group_config?.has_roles) {
+      return null;
+    }
+
+    return await getQuestionGroupPermissions(
+      instance_question?.id,
+      assessment_instance?.group_id,
+      user_id,
+    );
+  });
+
+  const panels: SubmissionPanels = {
     submissionPanel: null,
     extraHeadersHtml: null,
   };
 
-  // Fake locals. Yay!
-  const locals = {};
-  setLocalsFromConfig(locals);
-  Object.assign(
-    locals,
-    buildQuestionUrls(urlPrefix, variant, question, instance_question),
-    buildLocals(
+  const locals = {
+    ...buildQuestionUrls(urlPrefix, variant, question, instance_question),
+    ...buildLocals({
       variant,
       question,
       instance_question,
+      group_role_permissions,
       assessment,
       assessment_instance,
       assessment_question,
-    ),
-  );
+      group_config,
+    }),
+  };
+  setLocalsFromConfig(locals);
```

Comment:
Not sure if this is a bug or not, but this will override `locals.urlPrefix`, which could be different than the `urlPrefix` set by this function. I don't see this being used anywhere, since the only place where locals is used directly is in the question footer, which AFAICT doesn't use `locals.urlPrefix`, but I'm just raising this up as it may be a source of confusion in the future.

This is just an observation, I'm not asking for a change.

Reply:
Since this will now run as part of a normal HTTP request, I'm pretty sure `setLocalsFromConfig` can be removed entirely here.

Reply:
Not here, only in #10792...

Reply:
Doh, right, got my PRs mixed up.

Reply:
This is indeed a bug; it existed on master too. I was able to reproduce it by adding `<course-element></course-element>` to the `<pl-submission-panel>` on `exampleCourse/questions/demo/autograder/codeEditor` and then making a submission. When the submission results were rendered, it tried to load element assets from the wrong URL, which was visible as a 404 in the console and the network tab. I changed in 37400a1 to drop the call to `setLocalsFromConfig` entirely in favor of just explicitly setting `urlPrefix` and `plainUrlPrefix` on the `config` object here. I saw no such failed network requests after that.